core/schemas/graph.py

```diff
@@ -85,4 +86,61 @@ def load(cls, object: dict):
         return cls(**object)
 
 
-RelationshipTypes = Relationship | TagRelationship
+class Permission(IntFlag):
+    READ = 0b0001  # 1
+    WRITE = 0b0010  # 2
+    DELETE = 0b0100  # 4
+
+
+class Role:
+    NONE = Permission(0)
+    READER = Permission.READ
+    WRITER = Permission.READ | Permission.WRITE
+    OWNER = Permission.READ | Permission.WRITE | Permission.DELETE
+
+
+class RoleRelationship(BaseModel, database_arango.ArangoYetiConnector):
+    model_config = ConfigDict(str_strip_whitespace=True)
+    _exclude_overwrite: list[str] = list()
+    _collection_name: ClassVar[str] = "acls"
+    _type_filter: ClassVar[str | None] = None
+    _root_type: Literal["acl"] = "acl"
+    __id: str | None = None
+
+    source: str
+    target: str
+    role: Permission
+    created: datetime.datetime
+    modified: datetime.datetime
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        self.__id = data.get("__id", None)
+
+    # @computed_field(return_type=Literal["acl"])
+    @property
+    def root_type(self):
+        return self._root_type
+
+    # @computed_field(return_type=str)
+    @property
+    def id(self):
+        return self.__id
+
+    @classmethod
+    def load(cls, object: dict):
+        return cls(**object)
+
+    @classmethod
+    def has_permissions(cls, src, target_id: str, permission: Permission) -> bool:
+        vertices, paths, total = src.neighbors(
+            graph="acls", direction="outbound", max_hops=2
+        )
+        for path in paths:
+            for edge in path:
+                if edge.role & permission == permission and edge.target == target_id:
+                    return True
+        return False
+
+
+RelationshipTypes = Relationship | TagRelationship | RoleRelationship
```

Comment:
Should we keep these roles / permissions in graph? Don't you think it would be better to have them in a separated roles.py file?

Reply:
I moved some of it to a `roles.py` file